patches/server/0979-Add-whitelist-events.patch

```diff
@@ -0,0 +1,31 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: SageSphinx63920 <sage@sagesphinx63920.dev>
+Date: Sun, 14 May 2023 12:57:15 +0200
+Subject: [PATCH] Add whitelist events
+
+
+diff --git a/src/main/java/net/minecraft/server/players/UserWhiteList.java b/src/main/java/net/minecraft/server/players/UserWhiteList.java
+index a6c70decad1a496595944a5dd2aeeb387757fef3..c5ddfa915c93ed2608ee1a232628bfb5f6fd1a39 100644
+--- a/src/main/java/net/minecraft/server/players/UserWhiteList.java
++++ b/src/main/java/net/minecraft/server/players/UserWhiteList.java
+@@ -30,4 +30,20 @@ public class UserWhiteList extends StoredUserList<GameProfile, UserWhiteListEntr
+     protected String getKeyForUser(GameProfile gameProfile) {
+         return gameProfile.getId().toString();
+     }
++
++    //Paper start - Add whitelist events
++    @Override
++    public void add(UserWhiteListEntry entry) {
++        if (!new io.papermc.paper.event.player.PlayerWhitelistAddEvent(entry.getUser().getId()).callEvent()) return;
```

Comment:
This becomes kinda ugly for server owners imo.
Calling `/whitelist add <player>` will still result in a success message to the server admins even if the event was cancelled.

Obviously, this method sadly returns void so we cannot use pre-existing logic to provide feedback to the call if the removal succeeded, however imo this should be explored.

Reply:
So i may change the return to boolean but this would be some deeper change into the server code. Is this what you mean?

Reply:
Nah, we cannot change that methods return type without breaking things.
We could however, create a delegating method along the lines of 
```java
public boolean tryAdd(UserWhiteListEntry entry) {
    if (event.call().isCancelled()) return false;
    add(entry);
    return true;
}
```

and then update the Whitelist command to use said method.
Just an idea, I'd appreciate feedback for other people on this too. The current logic e.g. does also only log IO errors when editing the whitelist and still results in a success message, so its already not pretty. Hence this whole suggestion might be overkill.

Reply:
This could be done in the WhitelistCommand#L58

Reply:
Any more opinions here?